geode-core/src/main/java/org/apache/geode/cache/query/internal/index/AbstractIndex.java

```diff
@@ -2002,7 +2002,8 @@ void populateListForEquiJoin(List list, Object outerEntries, Object innerEntries
    */
   synchronized void setPdxStringFlag(Object key) {
     // For Null and Undefined keys do not set the isIndexedPdxKeysFlagSet flag
-    if (key == null || key == IndexManager.NULL || key == QueryService.UNDEFINED) {
+    if (isIndexedPdxKeysFlagSet || key == null || key == IndexManager.NULL
```

Comment:
Looks like you have two flags for the same thing... isIndexedPdxKeys and isIndexedPdxKeysFlagSet?